shared/src/price_estimate.rs

```diff
@@ -32,71 +32,58 @@ pub enum PriceEstimationError {
     /// buy_token (sell_amount / buy_amount).
-    /// amount must be nonzero.
-    async fn estimate_price(
-        &self,
-        sell_token: H160,
-        buy_token: H160,
-        amount: U256,
-        kind: OrderKind,
-    ) -> Result<BigRational, PriceEstimationError>;
-
-    /// Returns the expected gas cost for this given trade
-    /// amount must be nonzero.
-    async fn estimate_gas(
-        &self,
-        sell_token: H160,
-        buy_token: H160,
-        amount: U256,
-        kind: OrderKind,
-    ) -> Result<U256, PriceEstimationError>;
+    pub fn price_in_sell_token_rational(&self, query: &Query) -> Option<BigRational> {
+        let (sell_amount, buy_amount) = self.amounts(query);
+        amounts_to_price(sell_amount, buy_amount)
+    }
 
-    async fn estimate_price_as_f64(
-        &self,
-        sell_token: H160,
-        buy_token: H160,
-        amount: U256,
-        kind: OrderKind,
-    ) -> Result<f64, PriceEstimationError> {
-        self.estimate_price(sell_token, buy_token, amount, kind)
-            .await
-            .and_then(|price| {
-                price
-                    .to_f64()
-                    .ok_or_else(|| anyhow!("Cannot convert price ratio to float").into())
-            })
+    pub fn price_in_sell_token_f64(&self, query: &Query) -> f64 {
+        let (sell_amount, buy_amount) = self.amounts(query);
+        sell_amount.to_f64_lossy() / buy_amount.to_f64_lossy()
     }
+}
 
-    fn amount_to_estimate_prices_with(&self) -> U256;
+#[async_trait::async_trait]
+pub trait PriceEstimating: Send + Sync {
+    async fn estimate(&self, query: &Query) -> Result<Estimate, PriceEstimationError>;
 
     // Returns a vector of (rational) prices for the given tokens denominated
     // in denominator_token or an error in case there is an error computing any
     // of the prices in the vector.
-    async fn estimate_prices(
-        &self,
-        tokens: &[H160],
-        denominator_token: H160,
-    ) -> Vec<Result<BigRational, PriceEstimationError>> {
-        join_all(tokens.iter().map(|token| async move {
-            if *token != denominator_token {
-                self.estimate_price(
-                    denominator_token,
-                    *token,
-                    self.amount_to_estimate_prices_with(),
-                    OrderKind::Sell,
-                )
-                .await
-            } else {
-                Ok(num::one())
-            }
-        }))
-        .await
+    async fn estimates(&self, queries: &[Query]) -> Vec<Result<Estimate, PriceEstimationError>> {
+        // Naive default implementation that could be implemented more efficiently for some
+        // estimators.
+        join_all(queries.iter().map(|query| self.estimate(query))).await
     }
+
+    // Used when a price relative to the native token is needed without having a specific amount.
+    fn native_token_amount_to_estimate_prices_with(&self) -> U256;
```

Comment:
I was thinking about moving this out of the price estimator too because it is not its job to decide what a good native token amount is. Instead of passing it from the arguments into here it would be passed into the components that need a generic price without an amount. Not sure about this so decided to keep it for this PR.

Reply:
I kind of agree. I feel like the native token amount is a configuration detail/implementation detail that doesn't need to be leaked in the trait.

For example a "CoinGecko" or "ParaSwap" price estimator wouldn't care about this but still need to implement this method.